src/psyclone/gocean1p0.py

```diff
@@ -1541,139 +1520,622 @@ def resolve_argument_names(args):
         # Finally we can provide the updated argument list without name clashes
         sub.args = [kobj] + argument_names
 
-    def gen_data_on_ocl_device(self, parent):
+    def gen_ocl_buffers_initialisation(self, parent):
         # pylint: disable=too-many-locals
         '''
         Generate code to create data buffers on OpenCL device.
 
         :param parent: Parent subroutine in f2pygen AST of generated code.
         :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
         '''
-        grid_arg = self._arguments.find_grid_access()
+        # Get the root symbol table and the root f2pygen node
         symtab = self.root.symbol_table
-        # Ensure the fields required by this kernel are on device. We must
-        # create the buffers for them if they're not.
-        parent.add(UseGen(parent, name="fortcl", only=True,
-                          funcnames=["create_rw_buffer"]))
-        parent.add(CommentGen(parent, " Ensure field data is on device"))
+        module = parent
+        while module.parent:
+            module = module.parent
+
+        # Traverse all arguments and make sure the buffers are initialised
         for arg in self._arguments.args:
-            if arg.argument_type == "field" or \
-               (arg.argument_type == "grid_property" and not arg.is_scalar):
-                api_config = Config.get().api_conf("gocean1.0")
-                if arg.argument_type == "field":
-                    # fields have a 'data_on_device' property for keeping
-                    # track of whether they are on the device
-                    condition = ".NOT. {0}%data_on_device".format(arg.name)
-                    device_buff = "{0}%device_ptr".format(arg.name)
-                    host_buff = api_config.grid_properties["go_grid_data"]\
-                        .fortran.format(arg.name)
+            if arg.argument_type == "field":
+                # Get the init_buffer routine and insert a call for this field
+                init_buf = self.gen_ocl_initialise_buffer(module)
```

Comment:
I think this may be the source of the indentation problem - should `module` here be `parent` (which will be the IfThenGen node in the f2pygen AST)?